### PR TITLE
API rate limiting

### PR DESCRIPTION
**Pull request by 0xdiba**
After a long time :smile: ...
Solves #84 & #87

### CODE REVIEW

File: database/database.go [hunk 1]
```diff
@@ -71,6 +71,32 @@ func (db *DB) NewScan(domain string, rplay int) (Scan, error) {
...
+func (db *DB) GetLastScanTimeForTarget(target string) (int64, time.Time, error) {
+
+	var (
+		id int64
+		t  time.Time
+	)
+
+	row := db.QueryRow(`SELECT id, timestamp
+			    FROM scans
+			    WHERE target=$1 ORDER BY timestamp DESC LIMIT 1`, target)
```
Comment (jvehent) [inline database/database.go]: We probably need to index those columns

File: tlsobs-api/handlers.go [hunk 2]
```diff
@@ -44,9 +47,47 @@ func ScanHandler(w http.ResponseWriter, r *http.Request) {
...
+			log.WithFields(logrus.Fields{
+				"domain": domain,
+				"error":  err.Error(),
+			}).Error("Could not get last scan for target")
+			err = errors.New("Could not get last scan for target")
+			return
+		}
+
+		if previd != -1 { // check if previous scan exists
+			if time.Now().UTC().Sub(prevtime).Hours() <= 24 {
```
Comment (jvehent) [inline tlsobs-api/handlers.go]: let's make this 24 configurable

File: database/database.go (hunk 1 above)
Comment (0xdiba) [inline database/database.go]: They're going to be queried a lot so the probability is big...

File: tlsobs-api/handlers.go (hunk 2 above)
Comment (0xdiba) [inline tlsobs-api/handlers.go]: Cfg file, cmd flag or just code variable ? :smiling_imp:

Comment (jvehent) [inline tlsobs-api/handlers.go]: cfg file

File: tlsobs-api/handlers.go [hunk 3]
```diff
@@ -44,9 +49,47 @@ func ScanHandler(w http.ResponseWriter, r *http.Request) {
...
 	domain := r.FormValue("target")
-
 	if validateDomain(domain) {
 
+		rescan := false
+		if r.FormValue("rescan") == "true" {
+			rescan = true
+		}
+
+		previd, prevtime, err := db.GetLastScanTimeForTarget(domain)
```
Comment (jvehent) [inline tlsobs-api/handlers.go]: bypass this entirely if rescan==true, save time

Comment (jvehent) [inline tlsobs-api/handlers.go]: hmmm... it's needed for the 3 minutes hard limit, so maybe not...

File: tlsobs-api/handlers.go [hunk 4]
```diff
@@ -44,9 +49,47 @@ func ScanHandler(w http.ResponseWriter, r *http.Request) {
...
+					w.WriteHeader(http.StatusOK)
+					fmt.Fprint(w, resp)
+					return
+				}
+
+				// forced rescan has been requested
+				if time.Now().UTC().Sub(prevtime).Minutes() <= 3 { // rate limit scan requests for same target
+					if rescan {
+						w.WriteHeader(429) // 429 http status code is not exported ( https://codereview.appspot.com/7678043/ )
+						fmt.Fprint(w, "")
```
Comment (jvehent) [inline tlsobs-api/handlers.go]: What about returning `Last scan scheduled X seconds ago, wait another Y seconds to rescan` ?

File: tlsobs-api/handlers.go (hunk 3 above)
Comment (0xdiba) [inline tlsobs-api/handlers.go]: Yeah, rescan==true has to be rate limited too.

File: tlsobs-api/handlers.go (hunk 4 above)
Comment (0xdiba) [inline tlsobs-api/handlers.go]: Easy one :wink:

**Comment by jvehent** [comment]
looks going, going to deploy in stage, test, and merge if all goes well

**Comment by jvehent** [comment]
Works as expected. /me likes ! :+1: 

```
$ curl -X POST -i 'http://apistage201512031437-2sjrnmheum.elasticbeanstalk.com/api/v1/scan?target=public.etherpad-mozilla.org&rescan=true'
HTTP/1.1 429 Too Many Requests
Content-Type: text/plain; charset=utf-8
Date: Wed, 13 Jan 2016 21:20:29 GMT
Server: nginx/1.8.0
Content-Length: 114
Connection: keep-alive

Last scan for target public.etherpad-mozilla.org initiated 2m0.864744468s ago.
Please try again in 59.135255532s.
```

**Pull request merged**